### PR TITLE
`PeerManager` updates

### PR DESCRIPTION
**Pull request by pavel-kokolemin**
From @TheQuantumPhysicist 

> Connecting from RPC will need a little more work, because we want to make it possible to maintain connections with a specific peer. So it's a superset of what addnode does in bitcoin. Substrate/Parity has 3 types of peers for this.
There's an issue about this.

I looked at Substrate but did not find 3 types there. I only saw 2 types there: normal and reserved.
RPC now has two commands to add and remove "reserved" node addresses. Adding a "reserved" node address with RPC won't initiate an outbound connection until `heartbeat` is called, and removing it won't disconnect currently connected peers. Reserved nodes are not banned.
I have renamed `addnode` to `reserved_node` everywhere because it sounds a bit better.
I also added a command line option to add "normal" peer addresses (`boot_node` option) at startup. This can be used if something is wrong with the DNS seeds, for example.
I think this PR partially resolves https://github.com/mintlayer/mintlayer-core/issues/621.

### CODE REVIEW

File: dns_server/src/crawler_p2p/crawler/tests/mod.rs [hunk 1]
```diff
@@ -129,13 +129,13 @@ fn randomized(#[case] seed: Seed) {
         })
         .collect::<Vec<_>>();
 
-    let added_count = rng.gen_range(0..5);
-    let added_nodes = nodes.choose_multiple(&mut rng, added_count).cloned().collect();
+    let reserved_count = rng.gen_range(0..5);
```
Comment (TheQuantumPhysicist) [inline dns_server/src/crawler_p2p/crawler/tests/mod.rs:132]: It's always better to randomize this 5 and that 10 and create a model that covers them both with the relation between them. This applies to all magic numbers.

File: p2p/src/interface/p2p_interface_impl_delegation.rs [hunk 2]
```diff
@@ -42,4 +42,12 @@ impl<T: Deref<Target = dyn P2pInterface> + DerefMut<Target = dyn P2pInterface> +
     async fn get_connected_peers(&self) -> crate::Result<Vec<ConnectedPeer>> {
         self.deref().get_connected_peers().await
     }
+
+    async fn add_reserved_node(&mut self, addr: String) -> crate::Result<()> {
```
Comment (TheQuantumPhysicist) [inline p2p/src/interface/p2p_interface_impl_delegation.rs:46]: What's the point of the `connect` function if we have this?

File: p2p/src/peer_manager/peerdb/address_data/tests.rs [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+            let is_valid_transition = match transition {
+                AddressStateTransitionTo::Connected => !address_data.is_connected(),
+                AddressStateTransitionTo::Disconnected => address_data.is_connected(),
+                AddressStateTransitionTo::ConnectionFailed => !address_data.is_connected(),
+                AddressStateTransitionTo::SetReserved => true,
+                AddressStateTransitionTo::UnsetReserved => true,
+            };
+
+            if is_valid_transition {
```
Comment (TheQuantumPhysicist) [inline p2p/src/peer_manager/peerdb/address_data/tests.rs:53]: What if it's not valid? No asserts? Nothing?

File: p2p/src/peer_manager/peerdb/mod.rs [hunk 4]
```diff
@@ -125,30 +142,53 @@ where
...
-        connected_outbound: &BTreeSet<A>,
+        connected_outbound_count: usize,
     ) -> Vec<A> {
+        let now = Instant::now();
         let count = MAX_OUTBOUND_CONNECTIONS
             .saturating_sub(pending_outbound.len())
-            .saturating_sub(connected_outbound.len());
+            .saturating_sub(connected_outbound_count);
 
         // TODO: Ignore banned addresses
```
Comment (TheQuantumPhysicist) [inline p2p/src/peer_manager/peerdb/mod.rs:156]: That sounds like a too-important bit to ignore for later and just throw in as an "unimportant TODO".

**Review by TheQuantumPhysicist — APPROVED** [review]
Looks fine. I'm just worried that there are way too many TODOs.

File: p2p/src/interface/p2p_interface_impl_delegation.rs (hunk 2 above)
Comment (TheQuantumPhysicist) [reply p2p/src/interface/p2p_interface_impl_delegation.rs:46]: Now, thinking about it, connect can be useful to just give a "one shot"... but then we should make this distinction clear.

File: p2p/src/peer_manager/peerdb/address_data/mod.rs [hunk 5]
```diff
@@ -47,30 +46,47 @@ pub enum AddressState {
     },
 }
 
+#[derive(Copy, Clone, Debug)]
+// Update `ALL_TRANSITIONS` if a new transition is added!
```
Comment (stanislav-tkach) [inline p2p/src/peer_manager/peerdb/address_data/mod.rs:50]: [`EnumIter` from the `strum` crate](https://docs.rs/strum_macros/0.24.3/strum_macros/derive.EnumIter.html) can be used instead.

**Review by stanislav-tkach — APPROVED** [review]

File: dns_server/src/crawler_p2p/crawler/tests/mod.rs (hunk 1 above)
Comment (pavel-kokolemin) [reply dns_server/src/crawler_p2p/crawler/tests/mod.rs:132]: Sorry, I'm not sure what you mean. I just wanted to allow testing of corner cases and this should be covered here.

File: p2p/src/interface/p2p_interface_impl_delegation.rs (hunk 2 above)
Comment (pavel-kokolemin) [reply p2p/src/interface/p2p_interface_impl_delegation.rs:46]: I updated comments.

File: p2p/src/peer_manager/peerdb/address_data/tests.rs (hunk 3 above)
Comment (pavel-kokolemin) [reply p2p/src/peer_manager/peerdb/address_data/tests.rs:53]: Nothing, the check is only needed to generate sane randomized input.

File: p2p/src/peer_manager/peerdb/mod.rs (hunk 4 above)
Comment (pavel-kokolemin) [reply p2p/src/peer_manager/peerdb/mod.rs]: I added the check for banned addresses.

File: p2p/src/peer_manager/peerdb/address_data/mod.rs (hunk 5 above)
Comment (pavel-kokolemin) [reply p2p/src/peer_manager/peerdb/address_data/mod.rs:50]: Right, thanks. It's only for tests so I think we can keep it.

File: dns_server/src/crawler_p2p/crawler/tests/mod.rs (hunk 1 above)
Comment (TheQuantumPhysicist) [reply dns_server/src/crawler_p2p/crawler/tests/mod.rs:132]: Instead of 5, use a random number. Instead of 10, use a random number. Define the boundaries of the random number so it covers a wide range over time.

Comment (pavel-kokolemin) [reply dns_server/src/crawler_p2p/crawler/tests/mod.rs:132]: Could you please write how it should look? I think it would be easier for me to understand.

Comment (TheQuantumPhysicist) [reply dns_server/src/crawler_p2p/crawler/tests/mod.rs:132]: As agreed in the call. This is OK.

File: p2p/src/interface/p2p_interface_impl_delegation.rs (hunk 2 above)
Comment (pavel-kokolemin) [reply p2p/src/interface/p2p_interface_impl_delegation.rs:46]: Here are the updated comments:

```
    /// Try to connect to a remote node (just once).
    /// For persistent connections `add_reserved_node` should be used.
    #[method(name = "connect")]
    async fn connect(&self, addr: String) -> rpc::Result<()>;

    /// Add the address to the reserved nodes list.
    /// The node will try to keep connections open to all reserved peers.
    #[method(name = "add_reserved_node")]
    async fn add_reserved_node(&self, addr: String) -> rpc::Result<()>;

    /// Remove the address from the reserved nodes list.
    /// Existing connection to the peer is not closed.
    #[method(name = "remove_reserved_node")]
    async fn remove_reserved_node(&self, addr: String) -> rpc::Result<()>;
```
